src/combinatorics/simple_graph/ends/defs.lean

```diff
@@ -0,0 +1,254 @@
+/-
+Copyright (c) 2022 Anand Rao, Rémi Bottinelli. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anand Rao, Rémi Bottinelli
+-/
+import combinatorics.simple_graph.basic
+import combinatorics.simple_graph.connectivity
+import data.set_like.basic
+import category_theory.category.basic
+import category_theory.filtered
+import topology.category.Top.limits
+
+/-!
+# Ends
+
+This file contains a definition of the ends of a simple graph, as sections of the inverse system
+assigning, to each finite set of vertices, the connected components of its complement.
+-/
+
+universes u
+variables {V : Type u} (G : simple_graph V) (K L L' M : set V)
+
+open classical
+
+noncomputable theory
+local attribute [instance] prop_decidable
+
+namespace simple_graph
+
+section out
+
+/-- The graph induced by removing `K` -/
+@[reducible] def out := G.induce $ Kᶜ
+
+/-- Subsetship induces an obvious map on the induced graphs. -/
+@[reducible] def out_hom {K L} (h : K ⊆ L) : G.out L →g G.out K :=
+{ to_fun := λ ⟨v, hvK⟩, ⟨v, set.compl_subset_compl.mpr h hvK⟩,
+  map_rel' := by { rintros ⟨_, _⟩ ⟨_, _⟩, exact id } }
+
+lemma out_hom_refl (K) : G.out_hom (subset_refl K) = hom.id :=
+by { ext ⟨_, _⟩, refl, }
+
+lemma out_hom_trans {K L M} (h : K ⊆ L) (h' : L ⊆ M) :
+  G.out_hom (h.trans h') = (G.out_hom h).comp (G.out_hom h') :=
+by { ext ⟨_, _⟩, refl, }
+
+lemma out_hom_injective {K} {L} (h : K ⊆ L) : function.injective (G.out_hom h) :=
+by { rintros ⟨v, _⟩ ⟨w, _⟩ e,
+    simpa only [out_hom, subtype.mk_eq_mk, rel_hom.coe_fn_mk] using e, }
+
+end out
+
+/-- The components outside a given set of vertices `K` -/
+@[reducible] def comp_out := (G.out K).connected_component
+
+variables {G} {K L M}
+
+/-- The set of vertices of `G` making up the connected component `C` -/
+@[reducible, simp] def comp_out.supp (C : G.comp_out K) : set V :=
+{v : V | ∃ h : v ∈ Kᶜ, connected_component_mk (G.out K) ⟨v, h⟩ = C}
```

Comment:
At some point we had code for `set_like V (G.connected_component)` if `G` is a graph on `V`, but in this case we want to coerce components in `G.out K` to `set V`. Would that work with what you're thinking of?

Reply:
What I was thinking of was about how components yield a partition of `V`, along with stuff about each component as a subgraph and how they partition `G`.

Reply:
Yeah, we had [stuff related to that](https://github.com/bottine/mathlib/blob/89719db2bf26bf96d32db52156a44bde1b9b8866/src/combinatorics/simple_graph/ends/for_mathlib/connected.lean#L357) but I figured since it wasn't already included, there was maybe some moral reason this kind of thing wasn't done. 
Like, what's the reason
```lean
instance (a : Type*) (r : a -> a -> Prop) : set_like (quot r)
```
doesn't exist?

Reply:
Regarding connected components, it's mostly a development bandwidth problem. For example, there is material about this starting roughly around [here](https://github.com/leanprover-community/mathlib/blob/kmill_hamiltonian/src/hamiltonian.lean#L285), but I feel like it still needs some more development before it should be included.

Regarding `set_like`, reason 1 is that I doubt it's crossed anyone's mind to give `quot r` a set-like instance, and reason 2 might be that having this set-like instance could be contentious. I would be worried about it because it makes every element of a quotient be a set as well -- we tend to use `set_like` in cases where the things are meant to be sets, like subgroups, etc. You could try arguing for a `set_like (quot r)` instance on Zulip.